scripts/excelToJson.py

```diff
@@ -34,125 +42,124 @@ def keys_to_lower_camel_case(data: Union[Dict, List]) -> Union[Dict, List]:
     }
     return section_data
 
 
+def generate_process_function():
+    """
+    Generate a processing function for the excel document
+    """
+    def process(xls, output_dir):
+        """
+        The processing function reads each sheet from an excel file, applies data transformations,
+        and generates the json objects based on the provided manual
+        """
+        sheet_json_content = []
+        for sheet_name in xls.sheet_names:
+            try:
+                df = pd.read_excel(xls, skiprows=[0], sheet_name = sheet_name,
+                                   dtype={"Ref Calculation": str, "Section Number": str, "RS Number": str}).dropna(
+                    axis=1, how='all')
+                df = remove_unnamed_columns(df)
+                missing_cols = [col for col in EXPECTED_COLUMNS if col not in df.columns]
+                for col in missing_cols:
+                    df[col] = None
+
+                df[["Basic","Normal","Additional info"]] = df[["Basic","Normal","Additional info"]].map(
+                    lambda x: x.lower() if isinstance(x, str) else x).replace({'yes': True, 'no': False}).fillna(False)
+
+                sheet_json_content.extend([prepare_section_json(group, sheet_name)
+                    for _, group in df.groupby(['Section Number', 'Section Description'])])
+
+            except Exception as e:
+                logging.error(f"Failed to process sheet {sheet_name}: {e}", exc_info=True)
+                raise
+
+        try:
+            with open(Path(output_dir) / "iva-defects.json", 'w', encoding='utf-8') as f:
+                json.dump(sheet_json_content, f, indent=4, cls=CustomEncoder, ensure_ascii=False)
+
+        except Exception as e:
+            logging.error(f"Failed to write to json file: {e}", exc_info=True)
+            raise
+
+    return process
+
+def process_excel_file(file_path: str, output_dir: str, max_workers: int = 5) -> None:
+    """
+    Process an excel file and generate json file
+    """
+    Path(output_dir).mkdir(parents=True, exist_ok=True)
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+        process = generate_process_function()
+        future = executor.submit(process, pd.ExcelFile(file_path), output_dir)
+
+        try:
+            future.result()
+        except Exception as e:
+            logging.error(f"Error processing sheet: {e}", exc_info=True)
+
+    logging.info("JSON files generated in the output directory.")
+
+
 if __name__ == "__main__":
-    excel_file = 'CVS Approvals Defect Taxonomy - Master Copy 1101023.xlsx'
-    out_dir = 'output_json'
-    cat_map = {
-        "M1": "Cars",
-        "M2 M3": "Buses and Coaches",
-        "N1": "Vans and LGV",
-        "N2 N3": "HGV",
-        "O1-O4": "Trailers",
-        "MSVA": "Motorcycle"
-    }
+    out_dir = '../tests/resources'
 
     try:
-        process_excel_file(excel_file, out_dir, cat_map)
+        process_excel_file(EXCEL_FILE, out_dir)
```

Comment:
How are you verifying that the json is correct and formatted correctly? Producing **all** of the json's correctly inside that file? Processing mistakes could easily go unnoticed as the outputted json file is really big

Reply:
Leo threw together a checker to check for duplicates etc